src/main/java/minicraft/entity/Entity.java

```diff
@@ -132,74 +134,127 @@ public boolean move(int xd, int yd) {
+						if (e != this) {
+							if (!blocked) blocked = e.blocks(this);
+							if (e instanceof Player) {
+								touchedBy(e);
+							} else {
+								e.touchedBy(this);
+							}
+						}
+						wasInside.add(e); // Add entity into list.
+					}
+				}
+				if (blocked) break;
+				y += sgn; // Movement successful
+				successful = true;
+			}
+		} else { // X axis
+			hitBoxLeft = y - yr;
+			hitBoxRight = y + yr;
+			hitBoxFront = x + xr * sgn;
+			hitBoxFront1 = hitBoxFront + d;
+			hitBoxLeftTile = hitBoxLeft >> 4;
+			hitBoxRightTile = hitBoxRight >> 4;
+			hitBoxFrontTile = hitBoxFront >> 4;
+			hitBoxFrontTile1 = hitBoxFront1 >> 4;
+			maxFrontTile = hitBoxFrontTile1; // Value for full tile movement
+			// Skips the current tile by adding 1.
+			for (int front = hitBoxFrontTile + sgn; sgn < 0 ? front >= hitBoxFrontTile1 : front <= hitBoxFrontTile1; front += sgn) {
+				for (int horTile = hitBoxLeftTile; horTile <= hitBoxRightTile; horTile++) {
+					if (!level.getTile(front, horTile).mayPass(level, front, horTile, this)) {
+						maxFrontTile = front - sgn; // Rolls back a tile by subtracting 1.
+						break; // Tile hit box check stops.
+					}
+				}
 			}
-		}
 
-		isInside.removeAll(wasInside); // Remove all the entities that this one is already touching before moving.
-		for (Entity e : isInside) {
-			if (e == this) continue; // Can't interact with yourself
-			if (e.blocks(this)) return false; // If the entity prevents this one from movement, don't move.
+			maxFront = hitBoxFront +
+				(sgn > 0 ? Math.min(d, (maxFrontTile << 4) - hitBoxFront + (1 << 4) - 1) : Math.max(d, (maxFrontTile << 4) - hitBoxFront));
+			if (maxFront == hitBoxFront) return false; // No movement can be made.
+
+			// These lists are named as if the entity has already moved-- it hasn't, though.
+			HashSet<Entity> wasInside = new HashSet<>(level.getEntitiesInRect(getBounds())); // Gets all the entities that are inside this entity (aka: colliding) before moving.
+			for (int front = hitBoxFront; sgn < 0 ? front > maxFront : front < maxFront; front += sgn) {
+				boolean blocked = false; // If the entity prevents this one from movement, no movement.
+				for (Entity e : level.getEntitiesInRect(new Rectangle(x + sgn, y, xr * 2, yr * 2, Rectangle.CENTER_DIMS))) {
+					if (!wasInside.contains(e)) { // Skips entities that were touched.
+						if (e != this) {
+							if (!blocked) blocked = e.blocks(this);
+							if (e instanceof Player) {
+								touchedBy(e);
+							} else {
+								e.touchedBy(this);
+							}
+						}
+						wasInside.add(e); // Add entity into list.
+					}
+				}
+				if (blocked) break;
+				x += sgn; // Movement successful
+				successful = true;
+			}
 		}
 
-		// Finally, the entity moves!
-		x += xd;
-		y += yd;
-
-		return true; // the move was successful.
+		return successful;
 	}
 
 	/** Checks if the entity is able to naturally be despawned in general conditions. Handles (despawns) if true. */
```

Comment:
There must be a better way of doing this. Manually calculating the bounds and etc of the entity seemes a little bit strange to put in one single method, let alone the same class.

Reply:
Right, but the manual calculation is still needed to make sure these are working correctly. Hmm, probably rectangle interception could be involved to simplify boundary calculation.

Reply:
If the movement is extremely small here, the calculation is quite linear as loops are always skipped. Besides, if the movement is not small, looping calculation is always needed to search for hit box interactions.